In [1]:
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup
import requests
import time
import re

In [2]:
df=pd.read_csv("Bike.csv")
df

Bike Model  Model Year  \
0                  Honda Shine 125 Drum OBD2      2024.0   
1                          Aprilia Storm 125      2021.0   
2                       Honda Shine 125 Drum      2023.0   
3                        Yezdi Roadster Dark      2023.0   
4                             Aprilia SR 150      2018.0   
...                                      ...         ...   
1623                Jawa 42 Bobber Red Sheen      2023.0   
1624         Honda Hness Anniversary Edition      2022.0   
1625  Suzuki Avenis 125 Ride Connect Edition      2022.0   
1626                        Suzuki Gixxer SF      2016.0   
1627                   TVS Apache RTR 160 4V      2018.0   

                          Location  Price (Rs) Finance_Available(Yes/No)  
0                Pune, Maharashtra     95000.0                       Yes  
1       Dadar, Mumbai, Maharashtra     60000.0                       Yes  
2                Bokaro, Jharkhand     81000.0                       Yes  
3             Hyderabad, Telangana    165000.0                       Yes  
4                East Delhi, Delhi     61000.0                       Yes  
...                            ...         ...                       ...  
1623  Alandur, Chennai, Tamil Nadu    189000.0                       Yes  
1624  Dahisar, Mumbai, Maharashtra    175000.0                       Yes  
1625          Dehradun, Utarakhand     65000.0                       Yes  
1626  Bhiwandi, Thane, Maharashtra     42000.0                       Yes  
1627          Kolkata, West Bengal     62000.0                       Yes  

[1628 rows x 5 columns]

# Split Location column into City and State

In [3]:
df[['City', 'State']] = df['Location'].str.rsplit(",",n=1, expand=True)

In [4]:
df

Bike Model  Model Year  \
0                  Honda Shine 125 Drum OBD2      2024.0   
1                          Aprilia Storm 125      2021.0   
2                       Honda Shine 125 Drum      2023.0   
3                        Yezdi Roadster Dark      2023.0   
4                             Aprilia SR 150      2018.0   
...                                      ...         ...   
1623                Jawa 42 Bobber Red Sheen      2023.0   
1624         Honda Hness Anniversary Edition      2022.0   
1625  Suzuki Avenis 125 Ride Connect Edition      2022.0   
1626                        Suzuki Gixxer SF      2016.0   
1627                   TVS Apache RTR 160 4V      2018.0   

                          Location  Price (Rs) Finance_Available(Yes/No)  \
0                Pune, Maharashtra     95000.0                       Yes   
1       Dadar, Mumbai, Maharashtra     60000.0                       Yes   
2                Bokaro, Jharkhand     81000.0                       Yes   
3             Hyderabad, Telangana    165000.0                       Yes   
4                East Delhi, Delhi     61000.0                       Yes   
...                            ...         ...                       ...   
1623  Alandur, Chennai, Tamil Nadu    189000.0                       Yes   
1624  Dahisar, Mumbai, Maharashtra    175000.0                       Yes   
1625          Dehradun, Utarakhand     65000.0                       Yes   
1626  Bhiwandi, Thane, Maharashtra     42000.0                       Yes   
1627          Kolkata, West Bengal     62000.0                       Yes   

                  City         State  
0                 Pune   Maharashtra  
1        Dadar, Mumbai   Maharashtra  
2               Bokaro     Jharkhand  
3            Hyderabad     Telangana  
4           East Delhi         Delhi  
...                ...           ...  
1623  Alandur, Chennai    Tamil Nadu  
1624   Dahisar, Mumbai   Maharashtra  
1625          Dehradun    Utarakhand  
1626   Bhiwandi, Thane   Maharashtra  
1627           Kolkata   West Bengal  

[1628 rows x 7 columns]

In [5]:
# Drop the Duplicates 
df = df.drop_duplicates().reset_index(drop=True)

In [6]:
# Drop the Location column
df.drop(columns=['Location'], inplace=True)

In [7]:
df

Bike Model  Model Year  Price (Rs)  \
0                  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1                          Aprilia Storm 125      2021.0     60000.0   
2                       Honda Shine 125 Drum      2023.0     81000.0   
3                        Yezdi Roadster Dark      2023.0    165000.0   
4                             Aprilia SR 150      2018.0     61000.0   
...                                      ...         ...         ...   
1587                Jawa 42 Bobber Red Sheen      2023.0    189000.0   
1588         Honda Hness Anniversary Edition      2022.0    175000.0   
1589  Suzuki Avenis 125 Ride Connect Edition      2022.0     65000.0   
1590                        Suzuki Gixxer SF      2016.0     42000.0   
1591                   TVS Apache RTR 160 4V      2018.0     62000.0   

     Finance_Available(Yes/No)              City         State  
0                          Yes              Pune   Maharashtra  
1                          Yes     Dadar, Mumbai   Maharashtra  
2                          Yes            Bokaro     Jharkhand  
3                          Yes         Hyderabad     Telangana  
4                          Yes        East Delhi         Delhi  
...                        ...               ...           ...  
1587                       Yes  Alandur, Chennai    Tamil Nadu  
1588                       Yes   Dahisar, Mumbai   Maharashtra  
1589                       Yes          Dehradun    Utarakhand  
1590                       Yes   Bhiwandi, Thane   Maharashtra  
1591                       Yes           Kolkata   West Bengal  

[1592 rows x 6 columns]

In [8]:
# Find the null values in the dataframe
df.isna().sum()

Bike Model                   31
Model Year                   31
Price (Rs)                   39
Finance_Available(Yes/No)     0
City                         39
State                        63
dtype: int64

In [9]:
# Drop the null values
df= df.dropna()

In [10]:
# Checking is there ay null values are remaining
df.isna().sum()

Bike Model                   0
Model Year                   0
Price (Rs)                   0
Finance_Available(Yes/No)    0
City                         0
State                        0
dtype: int64

In [11]:
df

Bike Model  Model Year  Price (Rs)  \
0                  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1                          Aprilia Storm 125      2021.0     60000.0   
2                       Honda Shine 125 Drum      2023.0     81000.0   
3                        Yezdi Roadster Dark      2023.0    165000.0   
4                             Aprilia SR 150      2018.0     61000.0   
...                                      ...         ...         ...   
1587                Jawa 42 Bobber Red Sheen      2023.0    189000.0   
1588         Honda Hness Anniversary Edition      2022.0    175000.0   
1589  Suzuki Avenis 125 Ride Connect Edition      2022.0     65000.0   
1590                        Suzuki Gixxer SF      2016.0     42000.0   
1591                   TVS Apache RTR 160 4V      2018.0     62000.0   

     Finance_Available(Yes/No)              City         State  
0                          Yes              Pune   Maharashtra  
1                          Yes     Dadar, Mumbai   Maharashtra  
2                          Yes            Bokaro     Jharkhand  
3                          Yes         Hyderabad     Telangana  
4                          Yes        East Delhi         Delhi  
...                        ...               ...           ...  
1587                       Yes  Alandur, Chennai    Tamil Nadu  
1588                       Yes   Dahisar, Mumbai   Maharashtra  
1589                       Yes          Dehradun    Utarakhand  
1590                       Yes   Bhiwandi, Thane   Maharashtra  
1591                       Yes           Kolkata   West Bengal  

[1501 rows x 6 columns]

# Creating a Brand Column

In [12]:
df['Brand'] = df['Bike Model'].str.extract(r'^(\w+)')

C:\Users\Dell\AppData\Local\Temp\ipykernel_8712\43120325.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Brand'] = df['Bike Model'].str.extract(r'^(\w+)')


In [13]:
df

Bike Model  Model Year  Price (Rs)  \
0                  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1                          Aprilia Storm 125      2021.0     60000.0   
2                       Honda Shine 125 Drum      2023.0     81000.0   
3                        Yezdi Roadster Dark      2023.0    165000.0   
4                             Aprilia SR 150      2018.0     61000.0   
...                                      ...         ...         ...   
1587                Jawa 42 Bobber Red Sheen      2023.0    189000.0   
1588         Honda Hness Anniversary Edition      2022.0    175000.0   
1589  Suzuki Avenis 125 Ride Connect Edition      2022.0     65000.0   
1590                        Suzuki Gixxer SF      2016.0     42000.0   
1591                   TVS Apache RTR 160 4V      2018.0     62000.0   

     Finance_Available(Yes/No)              City         State    Brand  
0                          Yes              Pune   Maharashtra    Honda  
1                          Yes     Dadar, Mumbai   Maharashtra  Aprilia  
2                          Yes            Bokaro     Jharkhand    Honda  
3                          Yes         Hyderabad     Telangana    Yezdi  
4                          Yes        East Delhi         Delhi  Aprilia  
...                        ...               ...           ...      ...  
1587                       Yes  Alandur, Chennai    Tamil Nadu     Jawa  
1588                       Yes   Dahisar, Mumbai   Maharashtra    Honda  
1589                       Yes          Dehradun    Utarakhand   Suzuki  
1590                       Yes   Bhiwandi, Thane   Maharashtra   Suzuki  
1591                       Yes           Kolkata   West Bengal      TVS  

[1501 rows x 7 columns]

In [14]:
# Find the Unique in the column Company.
df['Brand'].unique()

array(['Honda', 'Aprilia', 'Yezdi', 'Suzuki', 'Yamaha', 'Bicycle',
       'Bajaj', 'TVS', 'KTM', 'Ola', 'BMW', 'Mach', 'Jawa', 'Royal',
       'Ampere', 'Ather', 'Hero', 'Zelio', 'Harley', 'Benelli',
       'Husqvarna', 'Fidato', 'Evtric', 'Joy', 'Vida', 'Triumph', 'BSA',
       'BGauss', 'Mahindra', 'Oreva', 'Vespa', 'KUOTA', 'Revolt', 'WaltX',
       'Ducati', 'Odysse', 'Okaya', 'QJ', 'Benling', 'Hyosung', 'Pure',
       'Kawasaki', 'Jitendra', 'River', 'E', 'iVOOMi', 'BattRE', 'Keeway',
       'Brazen', 'Oben', 'Kabira', 'KOMAKI', 'Okinawa', 'Simple', 'Arrow',
       'Deltic'], dtype=object)

In [15]:
# Create a dictionary for correcting brand names
brand_corrections = {
    "Royal": "Royal Enfield",
    "Mach": "Mach City",
    "E": "E-Motorad",
    "Bicycle": "Hero Cycles",
    "Joy": "Joy e-Bike",
    "Vida": "Vida (by Hero)",
    "BGauss": "BGAUSS",
    "Oreva": "Oreva Electric",
    "Revolt": "Revolt Motors",
    "Ola": "Ola Electric",
    "Ather": "Ather Energy",
    "Ampere": "Ampere Electric",
    "Okaya": "Okaya EV",
    "Oben": "Oben Electric",
    "Kabira": "Kabira Mobility",
    "Simple": "Simple Energy",
    "Jitendra": "Jitendra EV",
    "Benling": "Benling India",
    "Odysse": "Odysse Electric",
    "KOMAKI": "Komaki Electric",
    "Evtric": "Evtric Motors",
    "Fidato": "Fidato Evtech",
    "iVOOMi": "iVOOMi Energy",
    "River": "River Mobility",
    "Keeway": "Keeway Motors",
    "Arrow": "Arrow EV",
    "Brazen": "Brazen Motors",
    "Pure": "Pure EV",
    "Deltic": "Deltic EV"
}


In [17]:
# Apply the corrections to your DataFrame
df['Brand'] = df['Brand'].replace(brand_corrections)


C:\Users\Dell\AppData\Local\Temp\ipykernel_8712\2799078704.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Brand'] = df['Brand'].replace(brand_corrections)


In [19]:
# Brand column is created successfully
df['Brand'].unique()

array(['Honda', 'Aprilia', 'Yezdi', 'Suzuki', 'Yamaha', 'Hero Cycles',
       'Bajaj', 'TVS', 'KTM', 'Ola Electric', 'BMW', 'Mach City', 'Jawa',
       'Royal Enfield', 'Ampere Electric', 'Ather Energy', 'Hero',
       'Zelio', 'Harley', 'Benelli', 'Husqvarna', 'Fidato Evtech',
       'Evtric Motors', 'Joy e-Bike', 'Vida (by Hero)', 'Triumph', 'BSA',
       'BGAUSS', 'Mahindra', 'Oreva Electric', 'Vespa', 'KUOTA',
       'Revolt Motors', 'WaltX', 'Ducati', 'Odysse Electric', 'Okaya EV',
       'QJ', 'Benling India', 'Hyosung', 'Pure EV', 'Kawasaki',
       'Jitendra EV', 'River Mobility', 'E-Motorad', 'iVOOMi Energy',
       'BattRE', 'Keeway Motors', 'Brazen Motors', 'Oben Electric',
       'Kabira Mobility', 'Komaki Electric', 'Okinawa', 'Simple Energy',
       'Arrow EV', 'Deltic EV'], dtype=object)

In [20]:
df

Bike Model  Model Year  Price (Rs)  \
0                  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1                          Aprilia Storm 125      2021.0     60000.0   
2                       Honda Shine 125 Drum      2023.0     81000.0   
3                        Yezdi Roadster Dark      2023.0    165000.0   
4                             Aprilia SR 150      2018.0     61000.0   
...                                      ...         ...         ...   
1587                Jawa 42 Bobber Red Sheen      2023.0    189000.0   
1588         Honda Hness Anniversary Edition      2022.0    175000.0   
1589  Suzuki Avenis 125 Ride Connect Edition      2022.0     65000.0   
1590                        Suzuki Gixxer SF      2016.0     42000.0   
1591                   TVS Apache RTR 160 4V      2018.0     62000.0   

     Finance_Available(Yes/No)              City         State    Brand  
0                          Yes              Pune   Maharashtra    Honda  
1                          Yes     Dadar, Mumbai   Maharashtra  Aprilia  
2                          Yes            Bokaro     Jharkhand    Honda  
3                          Yes         Hyderabad     Telangana    Yezdi  
4                          Yes        East Delhi         Delhi  Aprilia  
...                        ...               ...           ...      ...  
1587                       Yes  Alandur, Chennai    Tamil Nadu     Jawa  
1588                       Yes   Dahisar, Mumbai   Maharashtra    Honda  
1589                       Yes          Dehradun    Utarakhand   Suzuki  
1590                       Yes   Bhiwandi, Thane   Maharashtra   Suzuki  
1591                       Yes           Kolkata   West Bengal      TVS  

[1501 rows x 7 columns]

In [21]:
 # Move the 'Brand' column to the beginning
cols = ['Brand'] + [col for col in df.columns if col != 'Brand']
df = df[cols]

In [22]:
df.head()

Brand                 Bike Model  Model Year  Price (Rs)  \
0    Honda  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1  Aprilia          Aprilia Storm 125      2021.0     60000.0   
2    Honda       Honda Shine 125 Drum      2023.0     81000.0   
3    Yezdi        Yezdi Roadster Dark      2023.0    165000.0   
4  Aprilia             Aprilia SR 150      2018.0     61000.0   

  Finance_Available(Yes/No)           City         State  
0                       Yes           Pune   Maharashtra  
1                       Yes  Dadar, Mumbai   Maharashtra  
2                       Yes         Bokaro     Jharkhand  
3                       Yes      Hyderabad     Telangana  
4                       Yes     East Delhi         Delhi

In [23]:
df.to_csv('new_Bike.csv')

In [25]:
df=pd.read_csv('new_Bike.csv')

In [27]:
brand_counts = df['Brand'].value_counts()
print(brand_counts)


Brand
Bajaj              233
Royal Enfield      225
Honda              176
Hero               158
TVS                156
Yamaha             147
Suzuki              60
Jawa                58
Ola Electric        56
KTM                 49
Yezdi               18
Triumph             18
BMW                 17
Aprilia             16
Harley              13
Revolt Motors       12
Ather Energy        12
Hero Cycles          9
Benelli              8
Vespa                5
Husqvarna            5
Vida (by Hero)       4
Ampere Electric      4
BSA                  3
Mahindra             3
Kawasaki             2
BGAUSS               2
Evtric Motors        2
Komaki Electric      2
Okaya EV             2
Keeway Motors        1
E-Motorad            1
iVOOMi Energy        1
BattRE               1
Joy e-Bike           1
Brazen Motors        1
River Mobility       1
Kabira Mobility      1
Okinawa              1
Simple Energy        1
Arrow EV             1
Oben Electric        1
Pure EV              1
Jiten

In [28]:
# Dictionary of bike brands with their country of origin
brand_origin = {
    'Honda': 'Japan',
    'Aprilia': 'Italy',
    'Yezdi': 'India',
    'Suzuki': 'Japan',
    'Yamaha': 'Japan',
    'Hero Cycles': 'India',
    'Bajaj': 'India',
    'TVS': 'India',
    'KTM': 'Austria',
    'Ola Electric': 'India',
    'BMW': 'Germany',
    'Mach City': 'India',
    'Jawa': 'Czech Republic',
    'Royal Enfield': 'India',
    'Ampere Electric': 'India',
    'Ather Energy': 'India',
    'Hero': 'India',
    'Zelio': 'India',
    'Harley': 'USA',
    'Benelli': 'Italy',
    'Husqvarna': 'Sweden',
    'Fidato Evtech': 'India',
    'Evtric Motors': 'India',
    'Joy e-Bike': 'India',
    'Vida (by Hero)': 'India',
    'Triumph': 'UK',
    'BSA': 'UK',
    'BGAUSS': 'India',
    'Mahindra': 'India',
    'Oreva Electric': 'India',
    'Vespa': 'Italy',
    'KUOTA': 'Italy',
    'Revolt Motors': 'India',
    'WaltX': 'India',
    'Ducati': 'Italy',
    'Odysse Electric': 'India',
    'Okaya EV': 'India',
    'QJ': 'China',
    'Benling India': 'China',
    'Hyosung': 'South Korea',
    'Pure EV': 'India',
    'Kawasaki': 'Japan',
    'Jitendra EV': 'India',
    'River Mobility': 'India',
    'E-Motorad': 'India',
    'iVOOMi Energy': 'India',
    'BattRE': 'India',
    'Keeway Motors': 'Hungary',
    'Brazen Motors': 'India',
    'Oben Electric': 'India',
    'Kabira Mobility': 'India',
    'Komaki Electric': 'India',
    'Okinawa': 'India',
    'Simple Energy': 'India',
    'Arrow EV': 'India',
    'Deltic EV': 'India'
}


In [31]:
# Map the origin to a new column
df['Brand_Origin'] = df['Brand'].map(brand_origin)

#  (Optional) Fill missing origins as 'Unknown'
df['Brand_Origin'] = df['Brand_Origin'].fillna('Unknown')

#  Check a few rows
print(df[['Brand', 'Brand_Origin']].head(20))

#  Check how many brands from each country
print(df['Brand_Origin'].value_counts())

           Brand Brand_Origin
0          Honda        Japan
1        Aprilia        Italy
2          Honda        Japan
3          Yezdi        India
4        Aprilia        Italy
5         Suzuki        Japan
6         Yamaha        Japan
7    Hero Cycles        India
8          Bajaj        India
9            TVS        India
10         Bajaj        India
11           TVS        India
12           KTM      Austria
13         Bajaj        India
14  Ola Electric        India
15         Bajaj        India
16   Hero Cycles        India
17           BMW      Germany
18     Mach City        India
19           KTM      Austria
Brand_Origin
India             918
Japan             385
Czech Republic     58
Austria            49
Italy              31
UK                 21
Germany            17
USA                13
Sweden              5
China               2
South Korea         1
Hungary             1
Name: count, dtype: int64


In [32]:
df.head()

Unnamed: 0    Brand                 Bike Model  Model Year  Price (Rs)  \
0           0    Honda  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1           1  Aprilia          Aprilia Storm 125      2021.0     60000.0   
2           2    Honda       Honda Shine 125 Drum      2023.0     81000.0   
3           3    Yezdi        Yezdi Roadster Dark      2023.0    165000.0   
4           4  Aprilia             Aprilia SR 150      2018.0     61000.0   

  Finance_Available(Yes/No)           City         State Brand_Origin  
0                       Yes           Pune   Maharashtra        Japan  
1                       Yes  Dadar, Mumbai   Maharashtra        Italy  
2                       Yes         Bokaro     Jharkhand        Japan  
3                       Yes      Hyderabad     Telangana        India  
4                       Yes     East Delhi         Delhi        Italy

In [33]:
# removing unnamed column
df = df.drop(columns=['Unnamed: 0'])


In [34]:
df.head()

Brand                 Bike Model  Model Year  Price (Rs)  \
0    Honda  Honda Shine 125 Drum OBD2      2024.0     95000.0   
1  Aprilia          Aprilia Storm 125      2021.0     60000.0   
2    Honda       Honda Shine 125 Drum      2023.0     81000.0   
3    Yezdi        Yezdi Roadster Dark      2023.0    165000.0   
4  Aprilia             Aprilia SR 150      2018.0     61000.0   

  Finance_Available(Yes/No)           City         State Brand_Origin  
0                       Yes           Pune   Maharashtra        Japan  
1                       Yes  Dadar, Mumbai   Maharashtra        Italy  
2                       Yes         Bokaro     Jharkhand        Japan  
3                       Yes      Hyderabad     Telangana        India  
4                       Yes     East Delhi         Delhi        Italy

In [36]:
def get_model_type(model):
    model = str(model).lower()
    if any(x in model for x in ['r15', 'duke', 'rc', 'speed', 'ninja', 'gs', 'rr', 'rs', 'race', 'yzf', 'apache rr']):
        return 'Sports'
    elif any(x in model for x in ['activa', 'jupiter', 'ntorq', 'pleasure', 'maestro', 'burgman', 'fascino', 'scooty', 'sxl', 'vxl', 'ray zr']):
        return 'Scooter'
    elif any(x in model for x in ['classic', 'meteor', 'bullet', 'avenger', 'roadster', 'bobber', 'imperiale', 'cruise']):
        return 'Cruiser'
    elif any(x in model for x in ['xpulse', 'himalayan', 'adventure', 'scram', 'ronin', 'v-strom']):
        return 'Adventure'
    elif any(x in model for x in ['electric', 's1', 'revolt', 'iqube', 'optima', 'ev', 'etron', 'battery', 'magnus', 'photon', 'komaki', 'faast']):
        return 'Electric'
    else:
        return 'Commuter'

df['Model_Type'] = df['Bike Model'].apply(get_model_type)


In [37]:
df[['Brand', 'Bike Model', 'Model_Type']].head(20)


Brand                       Bike Model Model_Type
0          Honda        Honda Shine 125 Drum OBD2   Commuter
1        Aprilia                Aprilia Storm 125   Commuter
2          Honda             Honda Shine 125 Drum   Commuter
3          Yezdi              Yezdi Roadster Dark    Cruiser
4        Aprilia                   Aprilia SR 150   Commuter
5         Suzuki                  Suzuki Lets 110   Commuter
6         Yamaha                    Yamaha R15 V4     Sports
7    Hero Cycles                Bicycle Decathlon   Commuter
8          Bajaj     Bajaj Avenger Cruise 220 BS6    Cruiser
9            TVS                         TVS Wego   Commuter
10         Bajaj  Bajaj Freedom 125 NG04 Disc LED   Commuter
11           TVS               TVS Apache RTR 180   Commuter
12           KTM                    KTM RC 200 GP     Sports
13         Bajaj           Bajaj Dominar 400 2025   Commuter
14  Ola Electric                     Ola S1X Plus   Electric
15         Bajaj              Bajaj Pulsar NS 160   Commuter
16   Hero Cycles                 Bicycle Hercules     Sports
17           BMW                      BMW G 310 R   Commuter
18     Mach City        Mach City Munich 21 Speed     Sports
19           KTM                     KTM Duke 390     Sports

In [38]:
df

Brand                              Bike Model  Model Year  Price (Rs)  \
0       Honda               Honda Shine 125 Drum OBD2      2024.0     95000.0   
1     Aprilia                       Aprilia Storm 125      2021.0     60000.0   
2       Honda                    Honda Shine 125 Drum      2023.0     81000.0   
3       Yezdi                     Yezdi Roadster Dark      2023.0    165000.0   
4     Aprilia                          Aprilia SR 150      2018.0     61000.0   
...       ...                                     ...         ...         ...   
1496     Jawa                Jawa 42 Bobber Red Sheen      2023.0    189000.0   
1497    Honda         Honda Hness Anniversary Edition      2022.0    175000.0   
1498   Suzuki  Suzuki Avenis 125 Ride Connect Edition      2022.0     65000.0   
1499   Suzuki                        Suzuki Gixxer SF      2016.0     42000.0   
1500      TVS                   TVS Apache RTR 160 4V      2018.0     62000.0   

     Finance_Available(Yes/No)              City         State  \
0                          Yes              Pune   Maharashtra   
1                          Yes     Dadar, Mumbai   Maharashtra   
2                          Yes            Bokaro     Jharkhand   
3                          Yes         Hyderabad     Telangana   
4                          Yes        East Delhi         Delhi   
...                        ...               ...           ...   
1496                       Yes  Alandur, Chennai    Tamil Nadu   
1497                       Yes   Dahisar, Mumbai   Maharashtra   
1498                       Yes          Dehradun    Utarakhand   
1499                       Yes   Bhiwandi, Thane   Maharashtra   
1500                       Yes           Kolkata   West Bengal   

        Brand_Origin Model_Type  
0              Japan   Commuter  
1              Italy   Commuter  
2              Japan   Commuter  
3              India    Cruiser  
4              Italy   Commuter  
...              ...        ...  
1496  Czech Republic    Cruiser  
1497           Japan     Sports  
1498           Japan   Commuter  
1499           Japan   Commuter  
1500           India   Commuter  

[1501 rows x 9 columns]

In [39]:
# Move "Finance_Available(Yes/No)" to the end of the DataFrame
col = 'Finance_Available(Yes/No)'

# Reorder columns
df = df[[c for c in df.columns if c != col] + [col]]


print(df.columns)


Index(['Brand', 'Bike Model', 'Model Year', 'Price (Rs)', 'City', 'State',
       'Brand_Origin', 'Model_Type', 'Finance_Available(Yes/No)'],
      dtype='object')


In [40]:
df.head()

Brand                 Bike Model  Model Year  Price (Rs)           City  \
0    Honda  Honda Shine 125 Drum OBD2      2024.0     95000.0           Pune   
1  Aprilia          Aprilia Storm 125      2021.0     60000.0  Dadar, Mumbai   
2    Honda       Honda Shine 125 Drum      2023.0     81000.0         Bokaro   
3    Yezdi        Yezdi Roadster Dark      2023.0    165000.0      Hyderabad   
4  Aprilia             Aprilia SR 150      2018.0     61000.0     East Delhi   

          State Brand_Origin Model_Type Finance_Available(Yes/No)  
0   Maharashtra        Japan   Commuter                       Yes  
1   Maharashtra        Italy   Commuter                       Yes  
2     Jharkhand        Japan   Commuter                       Yes  
3     Telangana        India    Cruiser                       Yes  
4         Delhi        Italy   Commuter                       Yes

In [42]:
df.dtypes

Brand                         object
Bike Model                    object
Model Year                   float64
Price (Rs)                   float64
City                          object
State                         object
Brand_Origin                  object
Model_Type                    object
Finance_Available(Yes/No)     object
dtype: object

In [44]:
# changing the data type of model year into string
df['Model Year'] = df['Model Year'].astype('Int64')
df['Model Year'] = df['Model Year'].astype('str')

In [45]:
df.dtypes

Brand                         object
Bike Model                    object
Model Year                    object
Price (Rs)                   float64
City                          object
State                         object
Brand_Origin                  object
Model_Type                    object
Finance_Available(Yes/No)     object
dtype: object

In [46]:
df['Price (Rs)'] = df['Price (Rs)'].astype('Int64')

In [47]:
df.dtypes

Brand                        object
Bike Model                   object
Model Year                   object
Price (Rs)                    Int64
City                         object
State                        object
Brand_Origin                 object
Model_Type                   object
Finance_Available(Yes/No)    object
dtype: object

In [48]:
df.isna().sum()

Brand                        0
Bike Model                   0
Model Year                   0
Price (Rs)                   0
City                         0
State                        0
Brand_Origin                 0
Model_Type                   0
Finance_Available(Yes/No)    0
dtype: int64

In [49]:
df.duplicated().sum()

0

In [50]:
df.nunique()

Brand                         56
Bike Model                   547
Model Year                    30
Price (Rs)                   209
City                         490
State                         25
Brand_Origin                  12
Model_Type                     6
Finance_Available(Yes/No)      2
dtype: int64

In [54]:
df['Finance_Available(Yes/No)'].value_counts()

Finance_Available(Yes/No)
Yes    1492
No        9
Name: count, dtype: int64

### Check the shape of the DataFrame

- The DataFrame is having 1501 rows and 9 columns.

In [55]:
df.shape

(1501, 9)

# Checking information about the data

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1501 entries, 0 to 1500
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Brand                      1501 non-null   object
 1   Bike Model                 1501 non-null   object
 2   Model Year                 1501 non-null   object
 3   Price (Rs)                 1501 non-null   Int64 
 4   City                       1501 non-null   object
 5   State                      1501 non-null   object
 6   Brand_Origin               1501 non-null   object
 7   Model_Type                 1501 non-null   object
 8   Finance_Available(Yes/No)  1501 non-null   object
dtypes: Int64(1), object(8)
memory usage: 107.1+ KB


## Checking the Stastical Description.

In [57]:
# For Numerical columns
df.describe()

Price (Rs)
count         1501.0
mean   108906.978681
std    107619.808548
min           2000.0
25%          45000.0
50%          80000.0
75%         140000.0
max        1500000.0

In [58]:
# For Object columns
df.describe(include = 'object')

Brand                 Bike Model Model Year       City         State  \
count    1501                       1501       1501       1501          1501   
unique     56                        547         30        490            25   
top     Bajaj  Royal Enfield Classic 350       2023  Bangalore   Maharashtra   
freq      233                         27        191         75           315   

       Brand_Origin Model_Type Finance_Available(Yes/No)  
count          1501       1501                      1501  
unique           12          6                         2  
top           India   Commuter                       Yes  
freq            918        837                      1492

## Checking for duplicate

In [59]:
df.duplicated().sum()

0

## Checking the null value

In [61]:
df.isna().sum()

Brand                        0
Bike Model                   0
Model Year                   0
Price (Rs)                   0
City                         0
State                        0
Brand_Origin                 0
Model_Type                   0
Finance_Available(Yes/No)    0
dtype: int64

 ## Checking for number of unique values in the data

In [62]:
df.nunique()

Brand                         56
Bike Model                   547
Model Year                    30
Price (Rs)                   209
City                         490
State                         25
Brand_Origin                  12
Model_Type                     6
Finance_Available(Yes/No)      2
dtype: int64

In [63]:
# Extract only the part after comma
df['City'] = df['City'].str.replace(r'^.*,\s*', '', regex=True)
df['City']

0             Pune
1           Mumbai
2           Bokaro
3        Hyderabad
4       East Delhi
           ...    
1496       Chennai
1497        Mumbai
1498      Dehradun
1499         Thane
1500       Kolkata
Name: City, Length: 1501, dtype: object

In [64]:
df

Brand                              Bike Model Model Year  Price (Rs)  \
0       Honda               Honda Shine 125 Drum OBD2       2024       95000   
1     Aprilia                       Aprilia Storm 125       2021       60000   
2       Honda                    Honda Shine 125 Drum       2023       81000   
3       Yezdi                     Yezdi Roadster Dark       2023      165000   
4     Aprilia                          Aprilia SR 150       2018       61000   
...       ...                                     ...        ...         ...   
1496     Jawa                Jawa 42 Bobber Red Sheen       2023      189000   
1497    Honda         Honda Hness Anniversary Edition       2022      175000   
1498   Suzuki  Suzuki Avenis 125 Ride Connect Edition       2022       65000   
1499   Suzuki                        Suzuki Gixxer SF       2016       42000   
1500      TVS                   TVS Apache RTR 160 4V       2018       62000   

            City         State    Brand_Origin Model_Type  \
0           Pune   Maharashtra           Japan   Commuter   
1         Mumbai   Maharashtra           Italy   Commuter   
2         Bokaro     Jharkhand           Japan   Commuter   
3      Hyderabad     Telangana           India    Cruiser   
4     East Delhi         Delhi           Italy   Commuter   
...          ...           ...             ...        ...   
1496     Chennai    Tamil Nadu  Czech Republic    Cruiser   
1497      Mumbai   Maharashtra           Japan     Sports   
1498    Dehradun    Utarakhand           Japan   Commuter   
1499       Thane   Maharashtra           Japan   Commuter   
1500     Kolkata   West Bengal           India   Commuter   

     Finance_Available(Yes/No)  
0                          Yes  
1                          Yes  
2                          Yes  
3                          Yes  
4                          Yes  
...                        ...  
1496                       Yes  
1497                       Yes  
1498                       Yes  
1499                       Yes  
1500                       Yes  

[1501 rows x 9 columns]

In [65]:
df.to_csv("Bike_3_final_for_visualization.csv",index = False)

In [66]:
model = df['Bike Model'].value_counts().head(10)

In [67]:
model

Bike Model
Royal Enfield Classic 350             27
Royal Enfield Thunderbird 350         21
Ola S1 Pro                            18
Honda Activa                          16
Bajaj Pulsar 150 DTSi                 16
Royal Enfield Hunter 350 Metro        15
Ola S1 Pro Gen2                       14
Bajaj Avenger Cruise 220              13
TVS Jupiter                           13
Royal Enfield Interceptor 650 Twin    12
Name: count, dtype: int64